libraries/rush-lib/src/logic/operations/BuildPlanPlugin.ts

```diff
@@ -0,0 +1,298 @@
+    hooks.beforeExecuteOperations.tapPromise(PLUGIN_NAME, createBuildPlan);
+
+    async function createBuildPlan(
+      recordByOperation: Map<Operation, IOperationExecutionResult>,
+      context: IExecuteOperationsContext
+    ): Promise<void> {
+      const { projectConfigurations, projectChangeAnalyzer } = context;
+      const disjointSet: DisjointSet<Operation> = new DisjointSet<Operation>();
+      const operations: IterableIterator<Operation> = recordByOperation.keys();
+      for (const operation of operations) {
+        disjointSet.add(operation);
+      }
+      const buildCacheByOperation: Map<Operation, IBuildPlanOperationCacheContext> = new Map<
+        Operation,
+        IBuildPlanOperationCacheContext
+      >();
+      for (const operation of operations) {
+        const { associatedProject, associatedPhase } = operation;
+        if (associatedProject && associatedPhase) {
+          const projectConfiguration: RushProjectConfiguration | undefined =
+            projectConfigurations.get(associatedProject);
+          const fileHashes: Map<string, string> | undefined =
+            await projectChangeAnalyzer._tryGetProjectDependenciesAsync(associatedProject, terminal);
+          const cacheDisabledReason: string | undefined = projectConfiguration
+            ? projectConfiguration.getCacheDisabledReason(fileHashes!.keys(), associatedPhase.name, operation)
+            : `Project does not have a ${RushConstants.rushProjectConfigFilename} configuration file, ` +
+              'or one provided by a rig, so it does not support caching.';
+          buildCacheByOperation.set(operation, { cacheDisabledReason });
+        }
+      }
+      clusterOperations(disjointSet, buildCacheByOperation);
+      const buildPlan: ICobuildPlan = createCobuildPlan(disjointSet, terminal, buildCacheByOperation);
+      logCobuildBuildPlan(buildPlan, terminal);
+    }
+  }
+}
+
+function generateCobuildPlanSummary(operations: Operation[], terminal: ITerminal): ICobuildPlan['summary'] {
+  const noOpOperations: Set<Operation> = new Set(operations.filter((e) => e.runner?.isNoOp));
+  const leafQueue: Operation[] = [];
+
+  for (const operation of operations) {
+    if (noOpOperations.has(operation)) {
+      continue;
+    }
+    let numberOfConsumers: number = 0;
+    for (const consumer of operation.consumers) {
+      if (!noOpOperations.has(consumer)) {
+        numberOfConsumers++;
+      }
+    }
+    if (numberOfConsumers === 0) {
+      leafQueue.push(operation);
+    }
+  }
+  let currentLeafNodes: Set<Operation> = new Set<Operation>();
+  const remainingOperations: Set<Operation> = new Set<Operation>(operations);
+  let depth: number = 0;
+  let maxWidth: number = leafQueue.filter((e) => !e.runner?.isNoOp).length;
+  const numberOfNodes: number[] = [maxWidth];
+  const depthToOperationsMap: Map<number, Set<Operation>> = new Map<number, Set<Operation>>();
+  depthToOperationsMap.set(depth, new Set(leafQueue));
+  do {
```

Comment:
This is presumably a standard well-known algorithm; could you put a comment with the name of the algorithm? Or a brief description of what it is supposed to be doing?

Reply:
Added in [a7cd9e7](https://github.com/microsoft/rushstack/pull/4660/commits/a7cd9e7fab410c96af178121fee0a13c28679486)